x-pack/legacy/plugins/ml/public/application/components/color_range_legend/color_range_legend.tsx

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useEffect, useRef, FC } from 'react';
+import d3 from 'd3';
+
+import { EuiText } from '@elastic/eui';
+
+const COLOR_RANGE_RESOLUTION = 10;
+
+interface ColorRangeLegendProps {
+  colorRange: (d: number) => string;
+  title?: string;
+  titleAlign?: 'left' | 'right';
+  width?: number;
+}
+
+/**
+ * Component to render a legend for color ranges to be used for color coding
+ * table cells and visualizations.
+ *
+ * @param props ColorRangeLegendProps
+ */
+export const ColorRangeLegend: FC<ColorRangeLegendProps> = ({
+  colorRange,
+  title,
+  titleAlign = 'left',
+  width = 300,
+}) => {
+  const d3Container = useRef<null | SVGSVGElement>(null);
+
+  const scale = d3.range(COLOR_RANGE_RESOLUTION + 1).map(d => ({
+    offset: (d / COLOR_RANGE_RESOLUTION) * 100,
+    stopColor: colorRange(d / COLOR_RANGE_RESOLUTION),
+  }));
+
+  useEffect(() => {
+    if (d3Container.current === null) {
+      return;
+    }
+
+    const wrapperHeight = 32;
+    const wrapperWidth = width;
+
+    const margin = { top: 2, bottom: 20, left: 8, right: 8 };
```

Comment:
Code LGTM overall. Just curious about how the specific numbers were determined for things like `margin` above and in calculating `legendAxis` below.

Reply:
This is now changed after some design feedback, it's now:

```ts
const margin = { top: 2, bottom: 20, left: justifyTicks ? 1 : 4, right: 1 };
```

- `top: 2` to adjust for the vertical alignment with the title text
- `bottom: 20` the margin is applied to the color gradient so this gives room for the axis ticks and labels
- `left/right: 1` necessary so the first and last axis tick doesn't get cut off
- `left: 4` when `justifyTicks === true` needs a larger margin so the first tick text doesn't get cut off (not enabled right now in this PR)